File: JavaAlgorithm/src/main/java/algo/tzashinorpu/Daily/COM/isRobotBounded_1041.java
package algo.tzashinorpu.Daily.COM;

public class isRobotBounded_1041 {
    public boolean isRobotBounded(String instructions) {
        int direction = 1; // 1:north 2:east 3:south 4:west
        int nsPosition = 0, wePosition = 0;
        char[] chars = instructions.toCharArray();
        for (char c : chars) {
            if (c == 'L') {
                if (direction == 1) {
                    direction = 5;
                }
                direction--;
            } else if (c == 'R') {
                if (direction == 4) {
                    direction = 0;
                }
                direction++;
            } else {
                if (direction == 1) {
                    nsPosition++;
                }
                if (direction == 3) {
                    nsPosition--;
                }
                if (direction == 2) {
                    wePosition++;
                }
                if (direction == 4) {
                    wePosition--;
                }
            }
        }
        // 还停留在原位置或方向未改变时有环
        if (nsPosition == 0 && wePosition == 0)
            return true;
        return direction != 1;
    }
}

File: JavascriptAlgorithm/zerotrac/1500_1600/isRobotBounded_1041.js
var isRobotBounded = function (instructions) {
    // 只有(x,y)不是原点，并且方向和原来的方向一致，最后才回不去
    let x = 0, y = 0, i = 0, dir = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    /*
             N (0)
               |
       W (3) ------ E (1)
               |
            S (2)
    */
    for (let j = 0; j < instructions.length; j++) {
        if (instructions[j] == 'R') {
            i = (i + 1) % 4;
        } else if (instructions[j] == 'L') {
            i = (i + 3) % 4;
        } else {
            x += dir[i][0];
            y += dir[i][1];
        }
    }
    return x == 0 && y == 0 || i > 0;
};